### ISSUE TITLE
[expr] The unclear point about whether it is UB for floating-pointer type

### ISSUE THREAD

**Issue by xmh0511** [cplusplus/CWG#77]
Full name of submitter (unless configured in github; will be published with the issue): Jim X

[expr.pre] p4 says
> If during the evaluation of an expression, **the result is not mathematically defined** or not in the range of representable values for its type, **the behavior is undefined**.

[expr.mul] p4
> If the second operand of / or % is zero the behavior is undefined.

The above two rules both specify the cases where UB can be touched. 

[expr.const] p5 bullet 8
> **an operation that would have undefined behavior** as specified in **[intro] through [cpp]**;

Consider this example: 
````cpp
#include <iostream>
#include <limits>

int  main(){
    if constexpr(std::numeric_limits<float>::is_iec559){  // ensure the following expression is evaluated when the condition is true
        //constexpr float v = 2.0f / 0;  // #1
        constexpr float v2 = std::numeric_limits<float>::quiet_NaN();  // #2
    }
}
````
GCC and Clang report that `#2` is ok but `#1` is ill-formed. However, "not a number" is not mathematically defined, which is considered UB, however, it is defined under ISO/IEC/IEEE 60559. According to [expr.mul] p4, `#1` is unconditionally UB, however, it is defined under ISO/IEC/IEEE 60559. 

The inconsistent thing here is that both `#1` and `#2` are defined under ISO/IEC/IEEE 60559 but `#1` is viewed as UB and `#2` is well-defined, according to the implementation. 

### Suggested resolution

Should we relax these conditions for floating-pointer type by saying them implementation-defined?

**Comment by frederick-vs-ja** [comment]
Related to CWG2168, and there is exactly a reflector thread: https://lists.isocpp.org/sg12/2021/09/0973.php.

I think GCC and Clang consider both `2.0f / 0` and `std::numeric_limits<float>::quiet_NaN()` to be well-defined, but the former one can't be a constant subexpression (like relational comparisons between pointers associated to different storage instances).

**Comment by xmh0511** [comment]
Actually, the operation `2.0f / 0` does not violate any other bullet in [expr.constant] other than the operation is specified to undefined behavior in [expr.mul]. The same point in such two cases is that the result of `2.0f / 0` is defined as positive infinity by IEEE 754.

**Comment by frederick-vs-ja** [comment]
> Actually, the operation `2.0f / 0` does not violate any other bullet in [expr.constant] other than the operation is specified to undefined behavior in [expr.mul]. The same point in such two cases is that the result of `2.0f / 0` is defined as positive infinity by IEEE 754.

IMO it is implied in CWG2168 that SG6's intent hasn't been reflected by the current requirements.

----

I guess a proper resolution should be explicitly conditionally allowing only LIA-1 classification and/or IEEE-754 rules to override the "mathematically-definedness". Simply making the conditions implementation-defined would allow implementation to define `2.0f / 0` to some finite number, which is likely unintended.

**Comment by eisenwave** [comment]
The current wording is defective because the evaluation of a function call expression or *id-expression* with lvalue-to-rvalue conversion may produce infinity, but SG6 wants `std::numeric_limits<T>::infinity()` to be okay, and compilers accept it in constant expressions. The goal should be to only make unary and binary arithmetic expressions undefined that yield NaN, not function call expressions, lambdas, etc.

@jensmaurer can you please amend CWG2168 with the following proposed resolution:

Change [[expr.pre] paragraph 4](https://eel.is/c++draft/expr.pre#4) as follows:

<blockquote>
<ins>An <i>arithmetic expression</i> is a unary plus or minus ([expr.unary.op]),
binary plus or minus ([expr.add]), or binary multiplication, division, or remainder ([expr.mul]) expression where every operands is of arithmetic type.</ins>
If <del>during the evaluation of</del> <ins>the result of</ins> an <ins>arithmetic</ins> expression,
<del>the result</del> is not mathematically defined or not in the range of representable values for its type, the behavior is undefined.<br><br>

<ins>[<i>Example:</i></ins>
```cpp
constexpr std::float32_t inf = std::numeric_limits<std::float32_t>::infinity(); // OK, infinity not produced by arithmetic expression
constexpr std::float32_t max = std::numeric_limits<std::float32_t>::max(); // OK
constexpr std::float32_t min = std::numeric_limits<std::float32_t>::min(); // OK
constexpr std::float32_t d0 = max / 0;     // error: division by zero ([expr.mul])
constexpr std::float32_t inf2 = max + max; // error: infinity produced
constexpr std::float32_t nan = inf - inf;  // error: NaN produced
constexpr std::float32_t zero = min / max;  // OK, result is too small to be represented, but is within the range of representable values
```
<ins>— <i>end example</i>]</ins>
</blockquote>

Delete [[basic.fundamental] paragraph 13](https://eel.is/c++draft/basic.fundamental#13):

<blockquote>
<del>The minimum range of representable values for a floating-point type is the most negative finite floating-point number representable in that type through the most positive finite floating-point number representable in that type. In addition, if negative infinity is representable in a type, the range of that type is extended to all negative real numbers; likewise, if positive infinity is representable in a type, the range of that type is extended to all positive real numbers.</del>

<br><del>[Note: Since negative and positive infinity are representable in ISO/IEC 60559 formats, all real numbers lie within the range of representable values of a floating-point type adhering to ISO/IEC 60559. — end note]</del>
</blockquote>

---

If we didn't delete paragraph 13, it would mean that `max + max` in the example is well-defined despite producing infinity. That is because the result (mathematically `max + max`) is in the range of representable values (all real numbers), even though the result is actually infinity. Honestly this is just very weird, adds ambiguity, and seems unmotivated. If we say that `x / 0` is UB despite being well-defined and producing infinity in IEC 60599, it seems unmotivated and inconsistent to allow `max + max` (which GCC rejects).

**Comment by jensmaurer** [comment]
[string.conversions] strtod refers to "range of representable values".  We're striking the definition above.   That seems not helpful.

**Comment by eisenwave** [comment]
Alright, let's try something else:

Change [[basic.fundamental] paragraph 13](https://eel.is/c++draft/basic.fundamental#13) as follows:

<blockquote>
The <del>minimum</del> range of representable values for a floating-point type is the most negative finite floating-point number representable in that type through the most positive finite floating-point number representable in that type. <del>In addition, if negative infinity is representable in a type, the range of that type is extended to all negative real numbers; likewise, if positive infinity is representable in a type, the range of that type is extended to all positive real numbers.</del>

<del><br>[Note: Since negative and positive infinity are representable in ISO/IEC 60559 formats, all real numbers lie within the range of representable values of a floating-point type adhering to ISO/IEC 60559. — end note]</del>
</blockquote>

When a value is too large for the C functions `strtof` etc., `HUGE_VAL` is returned, i.e. infinity. Defining the range of representable values to be all real numbers instead of the actually representable numbers seem pretty pointless then. I think the intent is that if `strtof` returns `HUGE_VAL`, we throw `out_of_range`, and this behavior is unaffected.

**Comment by jensmaurer** [comment]
Updated CWG2168

**Comment by eisenwave** [comment]
Thank you. It's worth noting that we have the long-standing problem that `is_iec559` can be `true` even though division by zero and various other operations are UB, and *maybe* we would like to have fully defined behavior in terms of ISO/IEC 60559 in that case, but maybe not during constant evaluation. However, all that would be evolutionary.

For now, I think there's much value in clarifying which parts we consider C++ UB (but possibly implemented in terms of ISO/IEC 60599), and which parts are actually well-defined. There's lots of implementation divergence in this area.

**Comment by eisenwave** [comment]
There has been extensive reflector discussion of CWG2168. People aren't happy with the current proposed resolution, which is seen as adding more UB into the language. There will be a paper in the December mailing which supersedes this issue; see

https://github.com/eisenwave/cpp-proposals/issues/89
https://isocpp.org/files/papers/D3899R0.html

**Comment by frederick-vs-ja** [comment]
IIUC CWG2168 should be considered resolved by P3899R3. However, there still seems an issue about `x / 0.0`.

**Comment by eisenwave** [comment]
Yeah, the core issue is resolved. The handling of `x / 0.0` is at least clearly UB now.

This will presumably be resolved by the floating-point annex, which SG6 has started working on in Brno 2026. I'm not sure you can guarantee anything about floating-point division by zero unless you at least know you're working with ISO/IEC 60559 types and operations.

**Comment by frederick-vs-ja** [comment]
> I'm not sure you can guarantee anything about floating-point division by zero unless you at least know you're working with ISO/IEC 60559 types and operations.

Maybe we can defer the discussion after the adoption of [P3938](https://wg21.link/p3938).
I _think_ the C++ standard can presume results of these sort of operations (in the ways similar to ISO/IEC 60559) when inf and NaN are more explicitly mentioned, and only render the operations UB if the presumed results can't be represented in the result type.
